aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py

```diff
@@ -234,16 +235,51 @@ def _put_record(self, data_record: DataRecord) -> None:
                     ":now_in_millis": {"N": str(int(now.timestamp() * 1000))},
                     ":inprogress": {"S": STATUS_CONSTANTS["INPROGRESS"]},
                 },
+                **(
+                    {"ReturnValuesOnConditionCheckFailure": "ALL_OLD"}  # type: ignore
+                    if self.boto3_supports_condition_check_failure(boto3.__version__)
+                    else {}
+                ),
             )
         except ClientError as exc:
             error_code = exc.response.get("Error", {}).get("Code")
             if error_code == "ConditionalCheckFailedException":
+                old_data_record = self._item_to_data_record(exc.response["Item"]) if "Item" in exc.response else None
+                if old_data_record is not None:
+                    logger.debug(
+                        f"Failed to put record for already existing idempotency key: "
+                        f"{data_record.idempotency_key} with status: {old_data_record.status}, "
+                        f"expiry_timestamp: {old_data_record.expiry_timestamp}, "
+                        f"and in_progress_expiry_timestamp: {old_data_record.in_progress_expiry_timestamp}",
+                    )
+                    self._save_to_cache(data_record=old_data_record)
+
+                    try:
+                        self._validate_hashed_payload(old_data_record=old_data_record, data_record=data_record)
+                    except IdempotencyValidationError as ive:
+                        raise ive from exc
+
+                    raise IdempotencyItemAlreadyExistsError(old_data_record=old_data_record) from exc
+
                 logger.debug(
                     f"Failed to put record for already existing idempotency key: {data_record.idempotency_key}",
                 )
-                raise IdempotencyItemAlreadyExistsError from exc
-            else:
-                raise
+                raise IdempotencyItemAlreadyExistsError() from exc
+
+            raise
+
+    @staticmethod
+    def boto3_supports_condition_check_failure(boto3_version: str):
+        version = boto3_version.split(".")
+        # Only supported in boto3 1.26.164 and above
+        if len(version) >= 3 and int(version[0]) == 1 and int(version[1]) == 26 and int(version[2]) >= 164:
+            return True
+        if len(version) >= 2 and int(version[0]) == 1 and int(version[1]) > 26:
+            return True
+        if int(version[0]) > 1:
+            return True
+
+        return False
```

Comment:
```suggestion
    @staticmethod
    def _boto3_supports_condition_check_failure() -> bool:
        """
        Check if the installed boto3 version supports condition check failure.

        Returns
        -------
        bool 
            True if the boto3 version supports condition check failure, False otherwise.
        """
        major, minor, patch = map(int, (boto3.__version__.split('.')))

        # Condition check failure support introduced in boto3 1.26.164
        if (major, minor) >= (1, 26) and (minor, patch) >= (26, 164):
            return True

        return False
```

We may remove the parameter and get the boto3 directly.

Reply:
After a conversation with @dastra, we updated the code:

```python
@staticmethod
def boto3_supports_condition_check_failure(boto3_version: str) -> bool:
    """
    Check if the installed boto3 version supports condition check failure.

    Params
    ------
    boto3_version: str
        The boto3 version

    Returns
    -------
    bool 
        True if the boto3 version supports condition check failure, False otherwise.
    """
    # Only supported in boto3 1.26.164 and above
    major, minor, *patch = map(int, boto3_version.split('.'))
    return (major, minor, *patch) >= (1, 26, 164)
```